buildSrc/src/main/java/org/elasticsearch/gradle/ReaperService.java

```diff
@@ -0,0 +1,136 @@
+    private Path inputDir;
+    private Path logFile;
+    private volatile Process reaperProcess;
+
+    public ReaperService(Logger logger, Path buildDir, Path inputDir) {
+        this.logger = logger;
+        this.buildDir = buildDir;
+        this.inputDir = inputDir;
+        this.logFile = inputDir.resolve("reaper.log");
+    }
+
+    /**
+     * Register a pid that will be killed by the reaper.
+     */
+    public void registerPid(String serviceId, long pid) {
+        String[] killPidCommand = OS.<String[]>conditional()
+            .onWindows(() -> new String[]{"Taskill", "/F", "/PID", String.valueOf(pid)})
+            .onUnix(() -> new String[]{"kill", "-9", String.valueOf(pid)})
+            .supply();
+        registerCommand(serviceId, killPidCommand);
+    }
+
+    /**
+     * Register a system command that will be run by the reaper.
+     */
+    public void registerCommand(String serviceId, String... command) {
+        ensureReaperStarted();
+
+        try {
+            Files.writeString(inputDir.resolve(serviceId + ".cmd"), String.join(" ", command));
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+    }
+
+    public void unregister(String serviceId) {
+        try {
+            Files.deleteIfExists(inputDir.resolve(serviceId + ".cmd"));
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+    }
+
+    void shutdown() {
+        if (reaperProcess != null) {
+            ensureReaperAlive();
+            try {
+                reaperProcess.getOutputStream().close();
+                logger.info("Waiting for reaper to exit normally");
+                if (reaperProcess.waitFor() != 0) {
+                    throw new GradleException("Reaper process failed. Check log at "
+                        + inputDir.resolve("error.log") + " for details");
+                }
+            } catch (Exception e) {
+                throw new RuntimeException(e);
+            }
+
+        }
+    }
+
+    private synchronized void ensureReaperStarted() {
+        if (reaperProcess == null) {
+            try {
+                // copy the reaper jar
+                Path jarPath = buildDir.resolve("reaper").resolve("reaper.jar");
+                Files.createDirectories(jarPath.getParent());
+                InputStream jarInput = ReaperPlugin.class.getResourceAsStream("/META-INF/reaper.jar");
+                try (OutputStream out = Files.newOutputStream(jarPath)) {
+                    jarInput.transferTo(out);
+                }
+
+                // ensure the input directory exists
+                Files.createDirectories(inputDir);
+
+                // start the reaper
+                ProcessBuilder builder = new ProcessBuilder(
+                    Jvm.current().getJavaExecutable().toString(), // same jvm as gradle
+                    "-Xms4m", "-Xmx16m", // no need for a big heap, just need to read some files and execute
```

Comment:
Does this really only use 16m of heap? Damn, Java's gotten lighter over the years.

Reply:
I can get a basic java program to work with 8m heap, but I thought 16 would be conservative. :)